### ISSUE TITLE
Improvement Suggestion: Accurate "True" Metrics on Validation Set

### ISSUE THREAD

**Issue by brge17** [keras-team/keras#8607]
Keras.metrics current implementation averages metrics in batchwise averages. I think there should be an option to cache the intermediate output and compute accurate metrics at the end of epoch for the entire validation set. This would unlock metrics such as AUC, Precision, Recall, etc...

The hacky solution that I have used thus far (included below):

1. Remove validation generator from fit_generator

2. Create a custom callback that takes a list of custom metric functions and the validation generator. At the end of the epoch hack the accurate metrics to the metrics log.

This is completely undesirable as it runs against the default Keras API. **I was hoping to make this an option in evaluate_generator.**

The callback (single metric) looks like (1.5x slower than predict on generator...):

```
class TrueMetrics(keras.callbacks.Callback):

    def __init__(self, gen_handle, func, name):
        ''' Constructor for True Metrics.

        :param gen_handle: Generator - data to compute metric.
        :param func: Function - metric(y_pred, y_true).
        :param mode: String - 'best' defined by 'min' or 'max'.
        :param name: String - metric name.
        '''
        self.gen_handle = gen_handle
        self.func = func
        self.name = name
        self.score = None

    def on_epoch_end(self, epoch, logs={}):
        ''' On epoch end compute the metrics on entire generator.

        :param epoch: Keras epoch - required input (unused).
        :param log: Keras logs - required input.
        '''

        # Batch and epoch params
        batch_size = self.gen_handle.batch_size
        n_samples = self.gen_handle.n_samples
        steps_per_epoch = self.gen_handle.steps_per_epoch

        # Create empty placeholder arrays
        output_shape = list(self.model.output_shape)

        # Replace 'None' with # Samples
        output_shape[0] = batch_size * steps_per_epoch
        y_true = np.squeeze(np.zeros(output_shape))
        y_pred = np.squeeze(np.zeros(output_shape))

        for ind, batch in enumerate(self.gen_handle.generate()):

            # Grab batch components
            X_batch = batch[0]
            y_true_batch = batch[1]

            # Compute the output on the batch
            y_pred_batch = self.model.predict_on_batch(X_batch)

            # Create ind
            start = ind * batch_size
            # Safeguard against improper sizing on last batch
            stop = start + y_pred_batch.shape[0]

            # Add the batch output to the placeholder arrays
            y_true[start:stop] = np.squeeze(y_true_batch)
            y_pred[start:stop] = np.squeeze(y_pred_batch)

            # Check for last batch
            if ind + 1 == steps_per_epoch:
                break

        # Trim batch padding
        y_true = y_true[:n_samples]
        y_pred = y_pred[:n_samples]

        # Compute metric
        self.score = self.func(y_true, y_pred)
        # Hack - write metric to logs for {ModelCheckpoint, EarlyStopping, etc...}
        logs[self.name] = np.float64(self.score)  #np.float64 for TensorBoard

```

**Note:** I am using a custom generator under the hood in this callback.

**Comment by SamuelMarks** [comment]
Any progress here?

Would especially be advantaged from efficient [statefully] optimisation on specificity & sensitivity.

**Comment by brge17** [comment]
Yes, you can use Stateful metrics.
